[AST] Flatten internal representation of SubstitutionMap.

DougGregor opened the pull request:
Collapse the out-of-line array of replacement types and the `DenseMap` used
for conformances into a single field that references out-of-line,
tail-allocated storage for all of the contents of a `SubstitutionMap`.
The conformances are now represented as a flat array whose indices
line up with the indices of the conformance requirements in the
underlying generic signtature.

Unique and `ASTContext`-allocate `SubstitutionMap` instances, making them
cheap to copy and opening the door for them to be stored directly in AST nodes.

DougGregor committed:
[AST] Reimplement SubstitutionMap::subst() in terms of getSubstitutionMap().

The primary implementation for substituting into substitution maps was
cloning the contents of the SubstitutionMap and then performing
in-place updates to the (new) data structure, which both had too many
dependencies on the data structure and subverted some of the checking
we get through the normal path.

Reimplement this function in terms of
GenericSignature::getSubstitutionMap(), which is the primary/canonical
way to form a substitution map.

DougGregor committed:
[AST] Internalize details of constructing a SubstitutionMap.

DougGregor committed:
[AST] Flatten internal representation of SubstitutionMap.

Collapse the out-of-line array of replacement types and the DenseMap used
for conformances into a single field that references out-of-line,
tail-allocated storage for all of the contents of a SubstitutionMap.
The conformances are now represented as a flat array whose indices
line up with the indices of the conformance requirements in the
underlying generic signtature.

Reference this storage via a single shared_ptr, so that copies of
SubstitutionMaps are cheap but we aren't (yet) uniqueing them or
storing them within the ASTContext.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test compiler performance

huonw commented on include/swift/AST/SubstitutionMap.h:
```diff
@@ -55,48 +44,150 @@ enum class CombineSubstitutionMaps {
+  ///
+  /// When null, this substitution map is empty, having neither a generic
+  /// signature nor any replacement types/conformances.
+  std::shared_ptr<Storage> storage;
```
Is this a `shared_ptr` because it's nice to be able to freely copy and `SubstitutionMap` are often transient so allocating in the global arena is going to create a lot of junk?

huonw replied:
Oh, wait, you preempted this question in your commit message. Cool.

huonw commented on lib/AST/SubstitutionMap.cpp:
```diff
@@ -99,6 +94,86 @@ bool SubstitutionMap::hasDynamicSelf() const {
+      auto conformances = sub.getConformances();
+      assert(reqts.size() == conformances.size());
+
+      for (unsigned i = 0, e = conformances.size(); i < e; i++) {
```
This could be `for (auto i : indices(conformances))`.

huonw commented on include/swift/AST/GenericSignature.h:
```diff
@@ -223,6 +223,17 @@ class alignas(1 << TypeAlignInBits) GenericSignature final
 
+  /// Compute the number of conformance requirements in this signature.
+  unsigned getNumConformanceRequirements() const {
+    unsigned result = 0;
```
This could be

```cpp
return llvm::count_if(getRequirements(),
    [&] (const Requirement &req) { return req.getKind() == RequirementKind::Conformance }));
```

although that's not entirely better.

DougGregor committed:
[Conformance checking] Skip potential witnesses with error types in them.

This fixes several crasher regressions from the flattening of
SubstitutionMap.

DougGregor committed:
[AST] Minor cleanups to SubstitutionMap.

DougGregor committed:
[AST] Unique and ASTContext-allocate SubstitutionMaps.

Prepare for SubstitutionMaps to be stored in other AST nodes by making
them ASTContext-allocated and uniqued (via a FoldingSet). They are now
cheap to copy and have trivial destructors.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented on include/swift/AST/SubstitutionMap.h:
```diff
@@ -55,48 +44,150 @@ enum class CombineSubstitutionMaps {
+/// to use for deriving nested types and conformances.
+///
+/// Depending on how the SubstitutionMap is constructed, the abstract types are
+/// either archetypes or interface types. Care must be exercised to only look
```
I'll update the comment in my next PR, thanks!
